instance/models/instance.py

```diff
@@ -446,29 +531,47 @@ def accepts_ssh_commands():
                 self.provision_mongo()
 
             # Provisioning (ansible)
-            server.mark_as_provisioning()
+            self.mark_as_provisioning()
             self.reset_ansible_settings(commit=True)
             log, exit_code = self.deploy()
             if exit_code != 0:
-                server.mark_provisioning_finished(success=False)
+                self.mark_provisioning_finished(success=False)
                 self.provision_failed_email(self.ProvisionMessages.PROVISION_ERROR, log)
                 return (server, log)
 
-            server.mark_provisioning_finished(success=True)
-
             # Reboot
             self.logger.info('Rebooting server %s...', server)
             server.reboot()
             server.sleep_until(accepts_ssh_commands)
             self.logger.info('Provisioning completed')
 
+            # Declare instance up and running
+            self.mark_provisioning_finished(success=True)
+
             return (server, log)
 
         except:
             self.server_set.terminate()
             self.provision_failed_email(self.ProvisionMessages.PROVISION_EXCEPTION)
             raise
 
+    @Instance.status.only_for(Status.WaitingForServer)
+    def mark_as_provisioning(self):
+        """
+        Indicate that this instance is being provisioned.
+        """
+        self._transition(self._status_to_configuring_server)
+
+    @Instance.status.only_for(Status.ConfiguringServer)
+    def mark_provisioning_finished(self, success):
+        """
+        Indicate that this instance is done provisioning, either due to success or failure.
+        """
+        if success:
+            self._transition(self._status_to_running)
+        else:
+            self._transition(self._status_to_configuration_failed)
```

Comment:
I think you can get rid of these methods (`mark_as_provisioning` and `mark_provisioning_finished`). The transitions already check that the current state is correct, so the `only_for` is redundant.

Instead of `self.mark_provisioning_finished(success=True)` you can just call `self._status_to_running()` directly (if you also remove `_transition` as I suggested).